### PR TITLE
Enabling Flexx plugins

### PR DESCRIPTION
**Pull request by almarklein**
*Don't merge this; this PR is only to provide an example*

This PR is meant to give an idea of what JLab plugins written in Flexx could look like, and what would be needed to make this work. There is much to think about and decide, I hope that this is a good starting point for a discussion.

## Overview

For the purpose of example, I added all code into JupyterLab. In reality, these should be separate packages (therefore, don't merge this).

* There is a `flexx_plugin_manager` module that should take care of most of the work.
* I modified JLab's `__init__.py` to make it load the JS and CSS assets that Flexx needs. It uses above module to make this as easy as possible.
* There's a package `myplugins` with a few modules that each implement one plugin. I tried to cover some aspects like making use of other services and registering commands.

## Trying it out

Just run `jupyter lab` on this branch. You'll need the latest master from Flexx.

Four plugin widgets are added to the dock area. They don't have a tab title and cannot be dragged around, because Flexx uses its own Phosphor, which makes some `typeof`'s fail. I'll be working on this next.

### How to let users register (and enable) the plugins that they wrote themselves or installed via pip/conda?

Should we have a similar (or the same) mechanism as [for common plugins](https://jupyterlab-tutorial.readthedocs.io/en/latest/labextensions.html)? 
```
jupyter labextension install path/to/my_extension my_extension
jupyter labextension enable my_extension [--sys-prefix]
```

### How to make Jlab load Flexx' assets?
Should we make changes to JLab, or do something like `jupyter serverextension enable --py jupyterlab_flexx`?

### How do people write a Flexx-based JLab plugin?

I think I've got this covered; a plugin must be a Flexx `Model` (or `Widget`), and have a `jlab_activate()` method and (optional) `JLAB_REQUIRES` class attribute to get access to services in the `jlab_activate()` method.
This makes it easy to transform any Flexx widget into a JLab plugin, and to develop/test such plugins outside of JLab.

### How to specify required services?

More of an implementation detail, but JLab requires that services are specified by tokens. This does not work very well in this case, so I made it use id strings, and look up the token corresponding to the id. The code that does this uses a private attribute though; this should be improved.

### CODE REVIEW

**Comment by blink1073** [comment]
Thanks for this @almarklein, I will look at this in earnest when I come back from holiday on Monday.

**Comment by blink1073** [comment]
I *think* we can make this work with a `serverextension` for `flexx`, a single `flexx` lab extension, and regular `flexx` lab extensions regular that depend on the base `flexx` extension, available as a cookie cutter.

**Comment by almarklein** [comment]
Links to docs on jupyter lab extensions:
* https://jupyterlab-tutorial.readthedocs.io/en/latest/labextensions.html
* https://github.com/jupyterlab/extension-builder
* https://github.com/jupyterlab/extension-cookiecutter-ts

Links to docs on jupyter server extensions:
* http://jupyter-notebook.readthedocs.io/en/latest/extending/handlers.html#writing-a-notebook-server-extension

**Comment by blink1073** [comment]
My notes from our conversation today:

- There should be a core `flexx` `serverextension` and a `labextension`.
- The `serverextension` should add lab-specific handlers and an endpoint to get a flex session id from a client
- The `labextension` should have the static assets needed for `flexx`
- Notionally, `flexx` will manage the loading of `flexx` plugins, including their JS and CSS.  We may decide to make these `labextensions` later.
- `flexx` should support importing JS modules, so that tokens can be imported instead of using ids

**Comment by blink1073** [comment]
@almarklein, what if we allowed Python-based lab extensions to provide additional data to `jupyter-config-data` in through a `_jupyter_labextension_config` function that would sit alongside `_jupyter_labextension_paths`?

**Comment by almarklein** [comment]
Another iteration ... the jupyterlab is almost free of changes:
* Adds a Jupyter server extension to register the Flexx handlers.
* Adds a LabExtension to define the Flexx "app", i.e. the model that manages the Flexx plugins. Sessions are created for each page load, and the session id is communicated to the page via `_jupyter_labextension_config`
* Wrote  `mondriaanplugin.py` to be a lab extension too. It's rather a stub with the sole purpose to make the server load the module so that the plugin can register itself with the flexx plugin manager.

Topics for further discussion:
* Needed a few tweaks to the Jupyterlab code for the extension handling. Some are bugs, I think. Another one is a hack to allow plugins without entry points (because `flexx-core.js` does not have one).
* Would be sweet if there was another hook in a lab extension to get a ref to the server application, so that we can register handlers there and get rid of the server extension.
* Using lab extensions to register plugins written in Flexx feeels awkward, because there is nothing to install (but we have to pretend that there is for the plugin to work). Can we allow plugins to be enabled without installing them? Or maybe flexx-based plugins should be registered via a different command?

**Comment by blink1073** [comment]
Notes from a video meeting today:

1. We need an optional `_jupyter_labextension_init` function that takes a web app
2. We need to apply the `jupyterlab/labapp.py` fixes from this PR
3. I will submit a PR against this PR to fix the flexx entry point
4. We will use separate commands for flexx plugins

File: jupyterlab/labapp.py [hunk 1]
```diff
@@ -101,6 +103,10 @@ def get(self):
                         css_files.append('%s/%s/%s' % (
                             EXTENSION_PREFIX, name, fname
                         ))
+                    elif os.path.splitext(fname)[1] == '.js':
```
Comment (blink1073) [inline jupyterlab/labapp.py]: This isn't needed when using a manifest.

**Comment by blink1073** [comment]
@almarklein, we decided against having the separate `_jupyter_labextension_init` function in today's JLab meeting, preferring instead to use the existing `serverextension` mechanism.

**Comment by almarklein** [comment]
Can I ask why? Not having it would mean users have to install both a server extension and a jlab extension.

**Comment by blink1073** [comment]
The reason being separation of concerns: making access to the server separate from access to the client.  This process will get easier once we clean up the extension handling mechanisms: https://github.com/jupyter/notebook/issues/1706.

**Comment by almarklein** [comment]
Ok, I made a first stab at a new package that provides the server extension, lab extension and CLI. I gave it an `enable` command that enables itself both as a server ext and lab ext, so the fact that there are two does not matter so much, in retrospect. It keeps track by itself of the flexx-based plugins and loads them dynamically. 

Code is here: https://github.com/zoofIO/flexxlab

Name is open for discussion. Also, should this live alongside Flexx' repo or inside the Jupyterlab org?

Closing this PR.

**Comment by blink1073** [comment]
Great!  That all sounds good, including the current location, since it is a non-standard lab extension (uses `window.jupyter.lab`).

**Pull request closed without merging**